### ISSUE TITLE
Focus lost when using AlpineJS $id after the Livewire request is sent

### ISSUE THREAD

**Issue by johanvanhelden** [alpinejs/alpine#3097]
### Discussed in https://github.com/alpinejs/alpine/discussions/2826

Also discussed in the Livewire repo: https://github.com/livewire/livewire/discussions/4790

<div type='discussions-op-text'>

<sup>Originally posted by **johanvanhelden** April 11, 2022</sup>
Hi all! Not sure if this is a bug, or me not understanding something.

I was watching Caleb's Laracon talk and he was showing of how to use Alpine to generate ids for elements.

So I just finished implementing it, but after I type some data into an input element, and the Livewire request is sent to the server, the input looses its focus.

**Demo:**
![LivewireAlpineDemo](https://user-images.githubusercontent.com/19389981/157751947-bbf80441-2eb7-4e46-a9d1-15b1fa32fa1c.gif)

The HTML for the input element looks like this:
```html
<div x-data="" x-id="['input']">
    <label class="block | text-sm font-medium text-gray-700 mb-2" x-bind:for="$id('input')" for="input-2">Name</label>
    <div>
        <input 
            x-bind:id="$id('input')" 
            type="text" 
            class="w-full block | shadow-sm rounded border-gray-300 sm:text-sm | focus:ring-blue-500 focus:border-blue-500" 
            wire:model="state.name" 
            autocomplete="name" 
            autofocus="autofocus" 
            id="input-2"
        >
    </div>
</div>
```

If I revert the `$id()` changes, everything works like it should, without anything loosing focus.

Unfortunately the ID stays the same after the async request. Before the reload it is `input-2` and after losing the focus it remains `input-2`. So that one does not change. I can actually not see any changes in the source. No part of the input group's HTML seems to change.

Also posted in the Livewire discussions section, but it's a bit dead in there. So maybe it's more Alpine's fault?
https://github.com/livewire/livewire/discussions/4790</div>

**Comment by marchershey** [comment]
I came here to report this. 

After spending about 4 hours trying to troubleshoot why my Stripe element was rerendering any time livewire sent a request, I was about to pull my hair out but then broke down each and every attribute in both Alpine and Livewire. Once I removed the "magical" `x-id` from my modal, everything worked. 

I knew I should have looked at the issues...

**Comment by devinvb** [comment]
@johanvanhelden I'm having the same issue unfortunately. Did you manage to find a solution?

**Comment by devinvb** [comment]
> > @johanvanhelden I'm having the same issue unfortunately. Did you manage to find a solution?
> 
> Unfortunately not.

It seems like adding `wire:key=""` to the input resolves this problem. However I'm not sure if having an empty wire key value might introduce bugs with DOM diffing..

**Comment by johanvanhelden** [comment]
> > > @johanvanhelden I'm having the same issue unfortunately. Did you manage to find a solution?
> > 
> > Unfortunately not.
> 
> It seems like adding `wire:key=""` to the input resolves this problem. However I'm not sure if having an empty wire key value might introduce bugs with DOM diffing..
> 
> 

Hmm, that does not sound like a good idea. If I understand things correctly, the key needs te be unique across the page due to how the DOM diffing works. So if you have multiple inputs, there will be duplicate empty keys.

**Comment by devinvb** [comment]
>
> Hmm, that does not sound like a good idea. If I understand things correctly, the key needs te be unique across the page due to how the DOM diffing works. So if you have multiple inputs, there will be duplicate empty keys.

Yes that's what I thought too. It seems to work however, also having multiple fields with the empty wire:key attribute in the same livewire component and conditionally showing one of these inputs works as expected.

Maybe @calebporzio can give some more information on this?

**Comment by TobiasFroberg** [comment]
I solved this by **not** using `x-id`, but instead using the longer more verbose method as [outlined in the docs](https://alpinejs.dev/magics/id#groups-with-x-id).

Pasting the code from the docs:
```html
<div x-data="{ id: $id('text-input') }">
    <label :for="id"> <!-- "text-input-1" -->
    <input type="text" :id="id"> <!-- "text-input-1" -->
</div>
 
<div x-data="{ id: $id('text-input') }">
    <label :for="id"> <!-- "text-input-2" -->
    <input type="text" :id="id"> <!-- "text-input-2" -->
</div>
```

**Comment by MichielBier** [comment]
Thanks @TobiasFroberg . Just did the same thing in my project, works like a charm!!
